File: sub-system-config/src/main/java/org/github/controller/ErrorController.java
package org.github.controller;

import java.util.List;
import java.util.Map;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import lombok.*;
import org.springframework.boot.autoconfigure.web.ServerProperties;
import org.springframework.boot.autoconfigure.web.servlet.error.BasicErrorController;
import org.springframework.boot.autoconfigure.web.servlet.error.ErrorViewResolver;
import org.springframework.boot.web.servlet.error.ErrorAttributes;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.servlet.ModelAndView;
import static cn.hutool.core.bean.BeanUtil.*;
import static org.github.spring.restful.json.JSONReturn.*;
import static org.springframework.http.HttpStatus.*;
import static org.springframework.http.ResponseEntity.ok;

@Controller
@RequestMapping("${server.error.path:${error.path:/error}}")
public class ErrorController extends BasicErrorController {
  public ErrorController(ErrorAttributes errorAttributes, ServerProperties serverProperties, List<ErrorViewResolver> errorViewResolvers) {
    super(errorAttributes, serverProperties.getError(), errorViewResolvers);
  }

  @Deprecated
  @Override
  public ModelAndView errorHtml(HttpServletRequest request, HttpServletResponse response) {
    return super.errorHtml(request, response);
  }

  @RequestMapping
  @Override
  public ResponseEntity<Map<String,Object>> error(HttpServletRequest request) {
    val error = super.error(request);
    if (error.getStatusCode().equals(NOT_FOUND)) return ok().body(beanToMap(path(null)));
    return ok().body(beanToMap(external(null)));
  }
}

File: sub-web/src/main/java/org/github/controller/ErrorController.kt
package org.github.controller

import org.apache.shiro.authz.AuthorizationException
import org.github.exception.ParamsErrorException
import org.github.ops.error
import org.github.ops.log
import org.github.spring.restful.json.JSONReturn.error
import org.github.spring.restful.json.JSONReturn.warn
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice

@RestControllerAdvice
class ErrorController {
  private val log = ErrorController::class.log

  @ExceptionHandler(Throwable::class)
  fun throwableHandler(e: Throwable) = error().withRetMsg(e.localizedMessage).also {log.error(e) {}}

  @ExceptionHandler(ParamsErrorException::class)
  fun paramsErrorExceptionHandler(e: ParamsErrorException) = warn().withRetMsg(e.localizedMessage).also {log.error(e) {}}

  @ExceptionHandler(AuthorizationException::class)
  fun authorizationExceptionHandler(e: AuthorizationException) = warn().withRetMsg("权限不足").also {log.error(e) {}}
}
